opt: relax assertion for lookup join columns

DrewKimball opened the pull request:
#### opt: relax assertion for lookup join columns

This commit relaxes the assertion that all columns projected by a
lookup join are "required". In particular, we can ignore "required"
columns from the right input for semi- and anti-joins, since they
have an implicit projection that removes right-input columns.

There is no release note because the error can only be encountered
during testing.

Fixes #120136

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129711)
<!-- Reviewable:end -->

mw5h reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129711#-:-O5Jea4u1RQFb7xzw8-Z:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
The commit that added this check mentions how these extra columns can cause issues with costing: https://github.com/cockroachdb/cockroach/commit/81f3dfa64d35170990d4d37e7819624fba3bd9f2

So it seems like this check would be nice to have, in an ideal world. Did you explore what it would take to correct the lookup-join columns, mentioned here: https://github.com/cockroachdb/cockroach/issues/120136#issuecomment-1986209808

DrewKimball committed:
opt: relax assertion for lookup join columns

This commit relaxes the assertion that all columns projected by a
lookup join are "required". In particular, we can ignore "required"
columns from the right input for semi- and anti-joins, since they
have an implicit projection that removes right-input columns.

There is no release note because the error can only be encountered
during testing.

Fixes #120136

Release note: None

DrewKimball commented:
>The commit that added this check mentions how these extra columns can cause issues with costing: 81f3dfa
>
>So it seems like this check would be nice to have, in an ideal world. Did you explore what it would take to correct the lookup-join columns, mentioned here: #120136 (comment)

I've updated this PR to relax the restriction for semi- and anti-joins, instead of removing it altogether.

mgartner commented on pkg/sql/opt/xform/join_funcs.go:
```diff
@@ -480,6 +480,18 @@ func (c *CustomFuncs) generateLookupJoinsImpl(
+		//idx := m.Metadata().Table(t.Table).Index(t.Index)
+		//for i := range t.KeyCols {
+		//	requiredCols.Add(t.Table.ColumnID(idx.Column(i).Ordinal()))
+		//}
```
Looks like this was mistakenly included.

DrewKimball replied:
My mistake, removed it.

mgartner reviewed (approved):
Thanks for fixing this!

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129711#-:-O8x90lJ9X1jhijQo1tb:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mw5h)



<!-- Sent from Reviewable.io -->

DrewKimball commented:
TFTRs!

bors r+
